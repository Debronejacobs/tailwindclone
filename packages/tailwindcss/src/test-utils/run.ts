import { optimizeCss } from '@tailwindcss/optimize'
import { compile } from '..'

export function compileCss(css: string, candidates: string[] = []) {
  return optimizeCss(compile(css).build(candidates)).trim()
}

export function run(candidates: string[]) {
  return optimizeCss(compile('@tailwind utilities;').build(candidates)).trim()
}
