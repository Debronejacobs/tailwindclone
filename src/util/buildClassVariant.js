import escapeClassName from './escapeClassName'

export default function buildClassVariant(className, variantName, variantsAsSuffix, separator) {
  if (variantsAsSuffix) {
    return `${className}${escapeClassName(separator)}${variantName}`
  }
  return `.${variantName}${escapeClassName(separator)}${className.slice(1)}`
}
