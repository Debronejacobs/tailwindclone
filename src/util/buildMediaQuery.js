import _ from 'lodash'

function isInvalidValue(value) {
  return !value || value === '' || value === 0 || value === '0'
}

export default function buildMediaQuery(screens) {
  if (isInvalidValue(screens)) {
    return undefined
  }

  if (_.isString(screens)) {
    screens = { min: screens }
  }

  if (!Array.isArray(screens)) {
    screens = [screens]
  }

  return _(screens)
    .map((screen) => {
      if (_.has(screen, 'raw')) {
        return screen.raw
      }

      return _(screen)
        .map((value, feature) => {
          feature = _.get(
            {
              min: 'min-width',
              max: 'max-width',
            },
            feature,
            feature
          )

          if (isInvalidValue(value)) {
            return undefined
          }

          return `(${feature}: ${value})`
        })
        .uniq()
        .filter((v) => v !== undefined)
        .join(' and ')
    })
    .join(', ')
}
