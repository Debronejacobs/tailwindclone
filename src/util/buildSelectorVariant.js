import parser from 'postcss-selector-parser'
import tap from 'lodash/tap'

export default (selector, variantName, separator, onError = () => {}) =>
  parser(selectors => {
    tap(selectors.first.filter(({ type }) => type === 'class').pop(), classSelector => {
      if (classSelector === undefined) {
        onError('Variant cannot be generated because selector contains no classes.')
        return
      }

      classSelector.value = `${variantName}${separator}${classSelector.value}`
    })
  }).processSync(selector)
