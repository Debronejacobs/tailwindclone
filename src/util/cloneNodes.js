import _ from 'lodash'

export default nodes => _.map(nodes, node => node.clone())
