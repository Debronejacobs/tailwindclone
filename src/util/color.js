import namedColors from 'color-name'

let HEX = /^#([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})?$/i
let SHORT_HEX = /^#([a-f\d])([a-f\d])([a-f\d])([a-f\d])?$/i
let VALUE = `(?:\\d+|\\d*\\.\\d+)%?`
let SEP = `(?:,\\s*|\\s+)`
let ALPHA_SEP = `\\s*[,/]\\s*`
let RGB_HSL = new RegExp(
  `^(rgb|hsl)a?\\(\\s*(${VALUE})${SEP}(${VALUE})${SEP}(${VALUE})(?:${ALPHA_SEP}(${VALUE}))?\\s*\\)$`
)

export function parseColor(value) {
  if (typeof value !== 'string') {
    return null
  }

  value = value.trim()

  if (value in namedColors) {
    return {
      mode: 'rgb',
      color: namedColors[value],
    }
  }

  let hex = value
    .replace(SHORT_HEX, (_, r, g, b, a) => ['#', r, r, g, g, b, b, a ? a + a : ''].join(''))
    .match(HEX)

  if (hex !== null) {
    return {
      mode: 'rgb',
      color: [parseInt(hex[1], 16), parseInt(hex[2], 16), parseInt(hex[3], 16)],
      alpha: hex[4] ? parseInt(hex[4], 16) / 255 : undefined,
    }
  }

  let match = value.match(RGB_HSL)

  if (match !== null) {
    return {
      mode: match[1],
      color: [match[2], match[3], match[4]],
      alpha: match[5],
    }
  }

  return null
}

export function formatColor({ mode, color, alpha }) {
  let hasAlpha = alpha !== undefined
  return `${mode}${hasAlpha ? 'a' : ''}(${color.join(', ')}${hasAlpha ? `, ${alpha}` : ''})`
}
