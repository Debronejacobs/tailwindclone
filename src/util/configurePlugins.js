export default (pluginConfig, plugins) => {
  const pluginNames = Array.isArray(pluginConfig)
    ? pluginConfig
    : Object.keys(plugins).filter(
        pluginName => pluginConfig !== false && pluginConfig[pluginName] !== false
      )

  return pluginNames.map(pluginName => plugins[pluginName]())
}
