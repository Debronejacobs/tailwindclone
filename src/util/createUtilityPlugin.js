import _ from 'lodash'
import fromPairs from 'lodash/fromPairs'
import toPairs from 'lodash/toPairs'
import castArray from 'lodash/castArray'

function className(classPrefix, key) {
  if (key === '') {
    return classPrefix
  }

  if (key.startsWith('-')) {
    return `-${classPrefix}${key}`
  }

  return `${classPrefix}-${key}`
}

export default function createUtilityPlugin(themeKey, utilityVariations, mapTheme = theme => theme) {
  return function({ e, addUtilities, variants, theme }) {
    const utilities = utilityVariations.map(([classPrefix, properties]) => {
      return fromPairs(
        toPairs(mapTheme(theme(themeKey))).map(([key, value]) => {
          return [
            `.${e(className(classPrefix, key))}`,
            fromPairs(castArray(properties).map(property => [property, value])),
          ]
        })
      )
    })
    return addUtilities(utilities, variants(themeKey))
  }
}

export function convertDefaultKey(theme) {
  return _.mapKeys(theme, (value, key) => key === 'default' ? '' : key)
}
