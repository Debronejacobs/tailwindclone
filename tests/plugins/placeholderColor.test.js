import { quickPluginTest } from '../util/run'

quickPluginTest('placeholderColor', {
  safelist: [
    // Arbitrary values
    'placeholder-[#0088cc]',
    'placeholder-[var(--my-value)]',
  ],
  theme: {
    colors: {
      transparent: 'transparent',
      black: '#000',
      red: {
        100: '#f00',
        200: '#f00',
      },
    },
  },
}).toMatchSnapshot()

// With `placeholderOpacity` enabled
quickPluginTest('placeholderColor', {
  safelist: [
    // Arbitrary values
    'placeholder-[#0088cc]',
    'placeholder-[var(--my-value)]',
  ],
  corePlugins: ['placeholderOpacity'],
  theme: {
    colors: {
      transparent: 'transparent',
      black: '#000',
      red: {
        100: '#f00',
        200: '#f00',
      },
    },
    opacity: {
      0: '0',
      50: '.5',
    },
  },
}).toMatchSnapshot()
