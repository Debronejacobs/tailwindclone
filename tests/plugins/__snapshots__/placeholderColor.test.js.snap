// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the 'placeholderColor' plugin 1`] = `
"
.placeholder-\\[\\#0088cc\\]::placeholder {
  color: #08c;
}

.placeholder-\\[var\\(--my-value\\)\\]::placeholder {
  color: var(--my-value);
}

.placeholder-black::placeholder {
  color: #000;
}

.placeholder-black\\/0::placeholder {
  color: #0000;
}

.placeholder-black\\/10::placeholder {
  color: #0000001a;
}

.placeholder-black\\/100::placeholder {
  color: #000;
}

.placeholder-black\\/15::placeholder {
  color: #00000026;
}

.placeholder-black\\/20::placeholder {
  color: #0003;
}

.placeholder-black\\/25::placeholder {
  color: #00000040;
}

.placeholder-black\\/30::placeholder {
  color: #0000004d;
}

.placeholder-black\\/35::placeholder {
  color: #00000059;
}

.placeholder-black\\/40::placeholder {
  color: #0006;
}

.placeholder-black\\/45::placeholder {
  color: #00000073;
}

.placeholder-black\\/5::placeholder {
  color: #0000000d;
}

.placeholder-black\\/50::placeholder {
  color: #00000080;
}

.placeholder-black\\/55::placeholder {
  color: #0000008c;
}

.placeholder-black\\/60::placeholder {
  color: #0009;
}

.placeholder-black\\/65::placeholder {
  color: #000000a6;
}

.placeholder-black\\/70::placeholder {
  color: #000000b3;
}

.placeholder-black\\/75::placeholder {
  color: #000000bf;
}

.placeholder-black\\/80::placeholder {
  color: #000c;
}

.placeholder-black\\/85::placeholder {
  color: #000000d9;
}

.placeholder-black\\/90::placeholder {
  color: #000000e6;
}

.placeholder-black\\/95::placeholder {
  color: #000000f2;
}

.placeholder-red-100::placeholder {
  color: red;
}

.placeholder-red-100\\/0::placeholder {
  color: #f000;
}

.placeholder-red-100\\/10::placeholder {
  color: #ff00001a;
}

.placeholder-red-100\\/100::placeholder {
  color: red;
}

.placeholder-red-100\\/15::placeholder {
  color: #ff000026;
}

.placeholder-red-100\\/20::placeholder {
  color: #f003;
}

.placeholder-red-100\\/25::placeholder {
  color: #ff000040;
}

.placeholder-red-100\\/30::placeholder {
  color: #ff00004d;
}

.placeholder-red-100\\/35::placeholder {
  color: #ff000059;
}

.placeholder-red-100\\/40::placeholder {
  color: #f006;
}

.placeholder-red-100\\/45::placeholder {
  color: #ff000073;
}

.placeholder-red-100\\/5::placeholder {
  color: #ff00000d;
}

.placeholder-red-100\\/50::placeholder {
  color: #ff000080;
}

.placeholder-red-100\\/55::placeholder {
  color: #ff00008c;
}

.placeholder-red-100\\/60::placeholder {
  color: #f009;
}

.placeholder-red-100\\/65::placeholder {
  color: #ff0000a6;
}

.placeholder-red-100\\/70::placeholder {
  color: #ff0000b3;
}

.placeholder-red-100\\/75::placeholder {
  color: #ff0000bf;
}

.placeholder-red-100\\/80::placeholder {
  color: #f00c;
}

.placeholder-red-100\\/85::placeholder {
  color: #ff0000d9;
}

.placeholder-red-100\\/90::placeholder {
  color: #ff0000e6;
}

.placeholder-red-100\\/95::placeholder {
  color: #ff0000f2;
}

.placeholder-red-200::placeholder {
  color: red;
}

.placeholder-red-200\\/0::placeholder {
  color: #f000;
}

.placeholder-red-200\\/10::placeholder {
  color: #ff00001a;
}

.placeholder-red-200\\/100::placeholder {
  color: red;
}

.placeholder-red-200\\/15::placeholder {
  color: #ff000026;
}

.placeholder-red-200\\/20::placeholder {
  color: #f003;
}

.placeholder-red-200\\/25::placeholder {
  color: #ff000040;
}

.placeholder-red-200\\/30::placeholder {
  color: #ff00004d;
}

.placeholder-red-200\\/35::placeholder {
  color: #ff000059;
}

.placeholder-red-200\\/40::placeholder {
  color: #f006;
}

.placeholder-red-200\\/45::placeholder {
  color: #ff000073;
}

.placeholder-red-200\\/5::placeholder {
  color: #ff00000d;
}

.placeholder-red-200\\/50::placeholder {
  color: #ff000080;
}

.placeholder-red-200\\/55::placeholder {
  color: #ff00008c;
}

.placeholder-red-200\\/60::placeholder {
  color: #f009;
}

.placeholder-red-200\\/65::placeholder {
  color: #ff0000a6;
}

.placeholder-red-200\\/70::placeholder {
  color: #ff0000b3;
}

.placeholder-red-200\\/75::placeholder {
  color: #ff0000bf;
}

.placeholder-red-200\\/80::placeholder {
  color: #f00c;
}

.placeholder-red-200\\/85::placeholder {
  color: #ff0000d9;
}

.placeholder-red-200\\/90::placeholder {
  color: #ff0000e6;
}

.placeholder-red-200\\/95::placeholder {
  color: #ff0000f2;
}

.placeholder-transparent::placeholder, .placeholder-transparent\\/0::placeholder {
  color: #0000;
}

.placeholder-transparent\\/10::placeholder {
  color: #0000001a;
}

.placeholder-transparent\\/100::placeholder {
  color: #000;
}

.placeholder-transparent\\/15::placeholder {
  color: #00000026;
}

.placeholder-transparent\\/20::placeholder {
  color: #0003;
}

.placeholder-transparent\\/25::placeholder {
  color: #00000040;
}

.placeholder-transparent\\/30::placeholder {
  color: #0000004d;
}

.placeholder-transparent\\/35::placeholder {
  color: #00000059;
}

.placeholder-transparent\\/40::placeholder {
  color: #0006;
}

.placeholder-transparent\\/45::placeholder {
  color: #00000073;
}

.placeholder-transparent\\/5::placeholder {
  color: #0000000d;
}

.placeholder-transparent\\/50::placeholder {
  color: #00000080;
}

.placeholder-transparent\\/55::placeholder {
  color: #0000008c;
}

.placeholder-transparent\\/60::placeholder {
  color: #0009;
}

.placeholder-transparent\\/65::placeholder {
  color: #000000a6;
}

.placeholder-transparent\\/70::placeholder {
  color: #000000b3;
}

.placeholder-transparent\\/75::placeholder {
  color: #000000bf;
}

.placeholder-transparent\\/80::placeholder {
  color: #000c;
}

.placeholder-transparent\\/85::placeholder {
  color: #000000d9;
}

.placeholder-transparent\\/90::placeholder {
  color: #000000e6;
}

.placeholder-transparent\\/95::placeholder {
  color: #000000f2;
}
"
`;

exports[`should test the 'placeholderColor' plugin 2`] = `
"
.placeholder-\\[\\#0088cc\\]::placeholder {
  --tw-placeholder-opacity: 1;
  color: rgb(0 136 204 / var(--tw-placeholder-opacity));
}

.placeholder-\\[var\\(--my-value\\)\\]::placeholder {
  color: var(--my-value);
}

.placeholder-black::placeholder {
  --tw-placeholder-opacity: 1;
  color: rgb(0 0 0 / var(--tw-placeholder-opacity));
}

.placeholder-black\\/0::placeholder {
  color: #0000;
}

.placeholder-black\\/50::placeholder {
  color: #00000080;
}

.placeholder-red-100::placeholder {
  --tw-placeholder-opacity: 1;
  color: rgb(255 0 0 / var(--tw-placeholder-opacity));
}

.placeholder-red-100\\/0::placeholder {
  color: #f000;
}

.placeholder-red-100\\/50::placeholder {
  color: #ff000080;
}

.placeholder-red-200::placeholder {
  --tw-placeholder-opacity: 1;
  color: rgb(255 0 0 / var(--tw-placeholder-opacity));
}

.placeholder-red-200\\/0::placeholder {
  color: #f000;
}

.placeholder-red-200\\/50::placeholder {
  color: #ff000080;
}

.placeholder-transparent::placeholder, .placeholder-transparent\\/0::placeholder {
  color: #0000;
}

.placeholder-transparent\\/50::placeholder {
  color: #00000080;
}

.placeholder-opacity-0::placeholder {
  --tw-placeholder-opacity: 0;
}

.placeholder-opacity-50::placeholder {
  --tw-placeholder-opacity: .5;
}
"
`;
